bash_replace_or_append fixes

maage opened the pull request:
#### Description:

At #10534 I had some gripes about ensuring these jinja functions perform as wanted.

Here I extend bats test suite for `bash_replace_or_append`. Update test script. And use some basic generic bats features instead of doing it by hand.

Handle append situation where config does have newline at EOF. Implement it as own macro `bash_ensure_nl_at_eof` as there might be other cases where this functionality helps. There is some cases where the issue has been noted and acted upon by just unconditionally adding `\n` always, not fixed here other than what was in `bash_replace_or_append`.

Drop extra logic about symlinks in `bash_replace_or_append` with `sed -i` and just always follow them. `--follow-symlinks` should be default where ever there is `sed -i`.

Add `LC_ALL=C` to `sed` in `bash_replace_or_append`. Something like this should be always default.

Use bats features like `setup` and `$BATS_TEST_TMPDIR` to simplify tests. Less code to maintain.

Extend and change `tests/unit/bash/execute_tests.sh`. Now it shows failed output by default and I hope it helps to diagnose problems faster. There is some options to help diagnose issues.

Run bats tests parallel. This is kind of test parallel over ctest parallel. When bats features are used as is now, there should not be regressions. This is somewhat forward looking change as when test cases accumulate, it will help more and more. For simple developement cycle you just edit jinja macro and / or test suite and then run ctest for `bash-unit-tests` and time savings are visible.

#### Rationale:

`bash_replace_or_append` should be used even more. And it can be used more if there is some assurance it works as expected. Expanded test suite greatly. There are some skipped tests, could be new features, but don't see them as bugs at this time. But I decided was good to document just to be show what limitations are.

#### Review Hints:

There could be regressions because functionality has changed when file does not have newline at EOF. Or if there is other locale than C.

Also there could be regressions because of old bash versions and such.

CTF regression does impact testing, fix at: https://github.com/ComplianceAsCode/content-test-filtering/pull/44

maage commented:
Debian and Ubuntu has old version of bats, v0.4.0 and v1.2.1 respectively I guess. Added patch to disable features if so.

vojtapolasek commented:
/ok-to-test

vojtapolasek reviewed (changes requested):
This is great improvement to the project, thank you very much. I will try to cooperate with @mildas  to get the CTF fixed. See my small comment.

maage commented:
I also changed in `rm` in tests where checking non existing files and such not to use `-f`. And I also noticed I had misplaces some new tests, and moved all new tests to one patch.

maage commented:
This seems unrelated,
```
  File "/home/runner/work/content/content/tests/ssg_test_suite/common.py", line 285, in get_product_context
    product_yaml['cmake_build_type'] = 'Debug'
TypeError: 'Product' object does not support item assignment
Error: Process completed with exit code 1.
```
I guess issue is that `Product` does not by design have `__setitem__`, but now at `ssg_test_suite.common.get_product_context` dict is replaced with `Product` if `product` and then we try to set some values. 

```
% (source .pyenv.sh;python)
Python 3.8.16 (default, Dec  7 2022, 00:00:00) 
[GCC 12.2.1 20221121 (Red Hat 12.2.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import ssg
>>> import ssg.products
>>> pp = ssg.products.Product('products/fedora/product.yml')
>>> pp['a']=1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'Product' object does not support item assignment
>>> h={}
>>> h.update(pp)
>>> h['a']=1
```

I created #10582

maage commented:
/hold

This test is wrong,
```
@test "bash_replace_or_append - Key contains * with remediation" {
```

maage commented:
Needed to skip the test above. I have bats-2 patchset on top of this, that handles more issues. But it requires more work and touches more code.

`execute_tests.sh` had a bug when jinja was bad, rc=0

`call_set_config_file_rsyslog` had a bug,

Some potential shellcheck issues.

And style changes in new code and moved new tests in one block too.

I can change this to Draft and work all in one swoop. But it will take time.

Main production benefit here is `bash_ensure_nl_at_eof` and bats test infra changes. There could be other avenues already to benefit from them.

/hold cancel

vojtapolasek reviewed (approved):
Hello @maage and thank you for additional changes. I see that automatus is failing on some rules. However, I think it is not related to your changes. I checked all rules suggested by content test filtering and I ran Automatus against RHEL 7, RHEL 8, RHEL 9 and Fedora... and there are no unexpected failures.

maage commented:
Thanks
